wellcomeml/ml/clustering.py

```diff
@@ -0,0 +1,321 @@
+            outputs, besides a cluster id, a list of keyword for each
+            cluster.
+            params(dict): Dictionary containing extra parameters for
+            embedding, reducer or clustering models. Example:
+            {'clustering': {'eps': 0.1}, 'embedding': {'ngram_range': (1,2)}}
+        """
+        self.embedding = embedding
+        self.vectorizer = Vectorizer(embedding=embedding,
+                                     **params.get('embedding', {}))
+        self.n_kw = n_kw
+
+        reducer_dispatcher = {
+            'umap': umap.UMAP,
+            'tsne': TSNE
+        }
+        self.reducer = reducer_dispatcher[reducer](**params.get('reducer', {}))
+
+        clustering_dispatcher = {
+            'dbscan': DBSCAN
+        }
+
+        self.clustering = clustering_dispatcher[clustering](
+            **params.get('clustering', {})
+        )
+
+        self.cluster_ids = None
+        self.cluster_names = None
+        self.cluster_kws = None
+
+    def fit(self, X, *_):
+        """
+        Fits all clusters in the pipeline
+
+        Args:
+            X: A list of texts
+
+        Returns:
+            A TextClustering object
+
+        """
+        self._fit_step(X, step='vectorizer')
+        self._fit_step(step='reducer')
+        self._fit_step(step='clustering')
+
+        if self.embedding == 'tf-idf' and self.n_kw:
+            self._find_keywords(self.embedded_points.toarray(), n_kw=self.n_kw)
+
+        return self
+
+    def _fit_step(self, X=None, step='vectorizer'):
+        """Internal function for partial fitting only a certain step"""
+        if step == 'vectorizer':
+            self.embedded_points = self.vectorizer.fit_transform(X)
+        elif step == 'reducer':
+            self.reduced_points = \
+                self.reducer.fit_transform(self.embedded_points)
+        elif step == 'clustering':
+            self.clustering.fit(self.reduced_points)
+            self.cluster_ids = self.clustering.labels_
+
+    def optimise(self, X, param_grid, n_cluster_range=None, max_noise=0.2,
+                 verbose=False):
+        """
+        Optimises the clustering silhouette based on a parameter grid,
+        a range on number of clusters and a range on noise.
+
+        It is customised to avoid re-fitting of intermediate
+        steps (vectorizer and reducer) more than necessary.
+
+        Args:
+            X: A list of texts to be clustered
+            param_grid: A parameter grid, example:
+
+            param_grid = {'reducer': {'min_dist': [0.0, 0.2], 'n_neighbors': [
+            2,3,5], 'metric': ['cosine', 'euclidean']},
+            'clustering': {'min_samples': [2, 5], 'eps': [0.5, 1, 1.5]}}
+
+            n_cluster_range (2-uple of ints): A 2-uple describing the
+            max and min number of clusters (e.g.: (10, 20)). If unset,
+            will just choose the best silhopuette
+
+            max_noise (float in [0,1]): The maximum fraction of points
```

Comment:
is this param only applicable to dbscan?

Reply:
Certain algorithms (`DBScan`, `Optics`, `HDBscan`) which are thing are referred to as `noise clustering` do not cluster everything, and assign -1 to some borderline points. For other algorithms, the requirement will be vacuously satisfied, as `noise = 0`.